# integration-mapping-pipeline
Snakemake pipeline and docker environment used to analyze unidirectional amplicon sequencing data generated in this study.

This pipeline has been optimized for the experiments that include donors with individual UMIs. It should also work for 
experiments that lack the donor UMIs, just ignore all metrics in the output files that include `UMI` in their column name.

### Setting up the working directory
This snakemake pipeline requires a very specific working directory configuration in order to run properly.

Your working directory should follow this file/folder format:
```
WORKDIR/
├── metadata.tsv
├── 00.blacklist/
│   └── hg38.bed
├── 00.donor_check/
│   └── donor_check.fna
├── 00.donor_map/
│   ├── sampleA.fasta
│   ├── sampleA.bed
│   ├── sampleA.attd.fasta
│   └── ...
├── 00.fastq/
│   ├── sampleA.R1.fq.gz
│   ├── sampleA.R2.fq.gz
│   ├── sampleB.R1.fq.gz
│   ├── sampleB.R2.fq.gz
│   └── ...
└── 00.genome/
    ├── hg38.fna
    ├── hg38.fna.amb
    ├── hg38.fna.ann
    ├── hg38.fna.bwt
    ├── hg38.fna.fai
    ├── hg38.fna.pac
    ├── hg38.fna.sa
    ├── hg38.exons.gff3
    ├── hg38.genes.gff3
    └── hg38.genome.txt

```
* `metadata.tsv` is a file that contains important information about the samples that will be analyzed, and looks like:
  ```
  sample	lsr	biorep	techrep	i7_stagger	primer_seq	umi_start	umi_end
  sampleA	LSR1	1	1	TCGA	TCGATCGAGGTTGCATTCGG	24	36
  sampleB	LSR1	1	2		TCGATCGAGGTTGCATTCGG	24	36
  sampleC	LSR2	1	1	CA	TCGATCGAGGTTGCATTCGG	24	36
  ...
  ```
  * `sample` column is the name of the sample that corresponds with the names in `00.donor_map/` and `00.fastq`
  * `lsr` column is the name of the LSR used in this experiment, which is just a label that is used to group samples at later steps.
  * `biorep` column is the name of the biological replicate used in this experiment, which is just a label that is used to group samples at later steps.
  * `techrep` column is the name of the technical replicate used in this experiment, which is just a label that is used to group samples at later steps.
  * `i7_stagger` column is the stagger sequence placed before the primer to increase library sequencing complexity and will be removed early on in the pipeline.
  * `primer_seq` column is the primer sequence used to amplify the insertion junctions.
  * `umi_start` is the 0-indexed start of the UMI on the plasmid, where the position 0 is the first nucleotide of the primer.
  * `umi_end` is the 1-indexed end of the UMI on the plasmid. `umi_end - umi_start = length of the umi`


* `00.blacklist` contains a regions in the hg38 human genome that will be blacklisted due to technical errors. This file is called `hg38.bed` and looks something like this:
  ```
  chr6	73519430	73521921	EF1A_PROMOTER
  ```


* `00.donor_check/` contains sequences that are primarily used to filter out cross-contaminated amplicons. A small percentage of reads can contaminated other samples in the same run, and these can often be filtered out based on which donor plasmid they align to.
  * `donor_check.fna` is a file that must be generated by the user for a specific run, and looks like this:
    ```
    >sampleA|sampleB
    GACGGGCACCGGAGCCCGGCTCGATCGAGGTTGCATTCGGCGGCAAAAAAAAAAAACGGCAGGCATTAAACACAGATAAACAGTTAATGTTATTTCATTACCATTAACTGTTTATCTGTGGGAGTTTAGGGCTCCCAAGGTGCAATCGTG
    >sampleC
    GACGGGCACCGGAGCCCGGCTCGATCGAGGTTGCATTCGGCGGCAAAAAAAAAAAACGGCAGGCATTAAACACAGATAAACAGTTAATGTTATTATATTACCATTAACTGTTTATCTGTGGGAGTTTAGGGCTCCCAAGGTGCAATCGTG
    ```
    Where a subsequence of each donor plasmid is provided as individual FASTA records. This subsequence runs from 20 bases before the donor primer sequence to 20 bases after the attD. Samples with identical donors should be included as a single record with the sample names in the record id and separated by a `|`. The UMI should be replaced with a run of `A` nucleotides.


* `00.donor_map/` contains plasmid sequences and important regions in BED format.
  * `sampleA.fasta` contains a single FASTA record corresponding to the nucleotide sequence of the donor plasmid for `sampleA`. The name of the record should be `>sampleA` and the sequence itself should be centered around the attD in the plasmid.
  * `sampleA.bed` is a BED file for `sampleA` that includes these regions that should follow standard BED format:
    ```
    sampleA	2799	2819	Primer	0	+
    sampleA	2823	2835	UMI	0	.
    sampleA	2839	2909	attD	0	.
    sampleA	2868	2881	Center	0	.
    sampleA	2873	2875	Core	0	.
    ```
  * `sampleA.attd.fasta` is a FASTA file that contains one record that matches only the attD sequence of the `sampleA` donor plasmid, like this:
    ```
    >C1plus:2839-2909
    AGGCATTAAACACAGATAAACAGTTAATGTTATTTCATTACCATTAACTGTTTATCTGTGGGAGTTTAGG
    ```
    

* `00.fastq/` contains raw FASTQ files generated from paired sequencing of the unidirectional amplicons, with samples named appropriately. Please follow the exact file format described above.

* `00.genome/` contains the following files associated with the reference genome:
  * `hg38.fna` is your reference genome in FASTA format. It should also be indexed using `BWA`.
  * `hg38.exons.gff3` is a GFF3 file of only the exons found in your reference genome, that looks something like this:
    ```
    chr1	HAVANA	exon	11869	12227	.	+	.	ID=exon:ENST00000456328.2:1;Parent=ENST00000456328.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;exon_number=1;exon_id=ENSE00002234944.1;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1
    chr1	HAVANA	exon	12010	12057	.	+	.	ID=exon:ENST00000450305.2:1;Parent=ENST00000450305.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000450305.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=transcribed_unprocessed_pseudogene;transcript_name=DDX11L1-201;exon_number=1;exon_id=ENSE00001948541.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:37102;ont=PGO:0000005,PGO:0000019;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000002844.2
    chr1	HAVANA	exon	12179	12227	.	+	.	ID=exon:ENST00000450305.2:2;Parent=ENST00000450305.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000450305.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=transcribed_unprocessed_pseudogene;transcript_name=DDX11L1-201;exon_number=2;exon_id=ENSE00001671638.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:37102;ont=PGO:0000005,PGO:0000019;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000002844.2
    chr1	HAVANA	exon	12613	12721	.	+	.	ID=exon:ENST00000456328.2:2;Parent=ENST00000456328.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;exon_number=2;exon_id=ENSE00003582793.1;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1
    chr1	HAVANA	exon	12613	12697	.	+	.	ID=exon:ENST00000450305.2:3;Parent=ENST00000450305.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000450305.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=transcribed_unprocessed_pseudogene;transcript_name=DDX11L1-201;exon_number=3;exon_id=ENSE00001758273.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:37102;ont=PGO:0000005,PGO:0000019;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000002844.2
    chr1	HAVANA	exon	12975	13052	.	+	.	ID=exon:ENST00000450305.2:4;Parent=ENST00000450305.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000450305.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=transcribed_unprocessed_pseudogene;transcript_name=DDX11L1-201;exon_number=4;exon_id=ENSE00001799933.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:37102;ont=PGO:0000005,PGO:0000019;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000002844.2
    chr1	HAVANA	exon	13221	14409	.	+	.	ID=exon:ENST00000456328.2:3;Parent=ENST00000456328.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;exon_number=3;exon_id=ENSE00002312635.1;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1
    chr1	HAVANA	exon	13221	13374	.	+	.	ID=exon:ENST00000450305.2:5;Parent=ENST00000450305.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000450305.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=transcribed_unprocessed_pseudogene;transcript_name=DDX11L1-201;exon_number=5;exon_id=ENSE00001746346.2;level=2;transcript_support_level=NA;hgnc_id=HGNC:37102;ont=PGO:0000005,PGO:0000019;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000002844.2
    chr1	HAVANA	exon	13453	13670	.	+	.	ID=exon:ENST00000450305.2:6;Parent=ENST00000450305.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000450305.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=transcribed_unprocessed_pseudogene;transcript_name=DDX11L1-201;exon_number=6;exon_id=ENSE00001863096.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:37102;ont=PGO:0000005,PGO:0000019;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000002844.2
    chr1	HAVANA	exon	14404	14501	.	-	.	ID=exon:ENST00000488147.1:11;Parent=ENST00000488147.1;gene_id=ENSG00000227232.5;transcript_id=ENST00000488147.1;gene_type=unprocessed_pseudogene;gene_name=WASH7P;transcript_type=unprocessed_pseudogene;transcript_name=WASH7P-201;exon_number=11;exon_id=ENSE00001843071.1;level=2;transcript_support_level=NA;hgnc_id=HGNC:38034;ont=PGO:0000005;tag=basic,Ensembl_canonical;havana_gene=OTTHUMG00000000958.1;havana_transcript=OTTHUMT00000002839.1
    ...
    ```
  * `hg38.genes.gff3` is a GFF3 file of only the genes found in your reference genome, that looks something like this:
    ```
    chr1	HAVANA	gene	11869	14409	.	+	.	ID=ENSG00000223972.5;gene_id=ENSG00000223972.5;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;level=2;hgnc_id=HGNC:37102;havana_gene=OTTHUMG00000000961.2
    chr1	HAVANA	gene	14404	29570	.	-	.	ID=ENSG00000227232.5;gene_id=ENSG00000227232.5;gene_type=unprocessed_pseudogene;gene_name=WASH7P;level=2;hgnc_id=HGNC:38034;havana_gene=OTTHUMG00000000958.1
    chr1	ENSEMBL	gene	17369	17436	.	-	.	ID=ENSG00000278267.1;gene_id=ENSG00000278267.1;gene_type=miRNA;gene_name=MIR6859-1;level=3;hgnc_id=HGNC:50039
    chr1	HAVANA	gene	29554	31109	.	+	.	ID=ENSG00000243485.5;gene_id=ENSG00000243485.5;gene_type=lncRNA;gene_name=MIR1302-2HG;level=2;hgnc_id=HGNC:52482;tag=ncRNA_host;havana_gene=OTTHUMG00000000959.2
    chr1	ENSEMBL	gene	30366	30503	.	+	.	ID=ENSG00000284332.1;gene_id=ENSG00000284332.1;gene_type=miRNA;gene_name=MIR1302-2;level=3;hgnc_id=HGNC:35294
    chr1	HAVANA	gene	34554	36081	.	-	.	ID=ENSG00000237613.2;gene_id=ENSG00000237613.2;gene_type=lncRNA;gene_name=FAM138A;level=2;hgnc_id=HGNC:32334;havana_gene=OTTHUMG00000000960.1
    chr1	HAVANA	gene	52473	53312	.	+	.	ID=ENSG00000268020.3;gene_id=ENSG00000268020.3;gene_type=unprocessed_pseudogene;gene_name=OR4G4P;level=2;hgnc_id=HGNC:14822;havana_gene=OTTHUMG00000185779.1
    chr1	HAVANA	gene	57598	64116	.	+	.	ID=ENSG00000240361.2;gene_id=ENSG00000240361.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=OR4G11P;level=2;hgnc_id=HGNC:31276;havana_gene=OTTHUMG00000001095.3
    chr1	HAVANA	gene	65419	71585	.	+	.	ID=ENSG00000186092.7;gene_id=ENSG00000186092.7;gene_type=protein_coding;gene_name=OR4F5;level=2;hgnc_id=HGNC:14825;havana_gene=OTTHUMG00000001094.4
    chr1	HAVANA	gene	89295	133723	.	-	.	ID=ENSG00000238009.6;gene_id=ENSG00000238009.6;gene_type=lncRNA;gene_name=RP11-34P13.7;level=2;tag=overlapping_locus;havana_gene=OTTHUMG00000001096.2
    ...
    ```
  * `hg38.genome.txt` is a file listing the lengths of all of the records in your reference genome FASTA file, and looks something like this:
    ```
    chr1	248956422
    chr2	242193529
    chr3	198295559
    chr4	190214555
    ...
    ```
    
### Setting up the docker environment and running the pipeline
* Navigate to this subdirectory of this repository on your machine:
  ```
  cd /path/to/SystematicDiscoveryRecombinases2022/integration-mapping-pipeline/
  ```
  Next, run: 
  ```
  source setup.sh
  ```
  This step will have to be repeated whenever logging back into the machine, unless you automate it.

  If you haven't built the docker image, you can now build it with:
  ```
  integration_mapping_docker_build
  ```
  Now that the docker image is built, you can run your snakemake pipeline on your properly configured working directory with the command:
  ```
  integration_mapping_docker_snakemake WORKDIR/ 8
  ```
  Where `8` is a placehodler for the number of threads you'd like to use when executing the pipeline.

### Results folder
* Many directories will be created while the pipeline is executing. Here, we focus only on the `11.results` output folder.
* This is what the file architecture for this folder looks like:
  ```
  11.results/
  ├── all_integration_sites.fna
  ├── donor_umi_counts.tsv
  ├── merged_counts.sample.tsv
  ├── merged_counts.biorep.tsv
  ├── merged_counts.lsr.tsv
  ├── raw_counts.unfiltered.tsv
  ├── raw_counts.filtered.tsv
  ├── site_annotations.tsv
  ├── seqlogos/
  └── seqlogos.txt
  ```

* `all_integration_sites.fna` contains all integration sites detected across all the samples, and it looks like this:
  ```
  >chr1:630797-630799(+)
  TGCAACAGCTAAGGACTGCAAAACCCCACTCTGCATCAACTGAACGCAAATCAGCCACTTTAATTAAGCTAAGCCCTTACTAGACCAATGGGACTTAAACCC
  >chr1:1477579-1477581(+)
  CGCAGAGCCGCCCGAGAGGGAGGGTCAGTGTTGGTGAGGGCGTCTGGTCGTCCTGAGGGAGGGCCGGTGTTGGTGAGGGCATCTGGTCGTCCTGAGGGAGGG
  >chr1:1860108-1860110(+)
  aaaaaaagtaaacaaagaagaagtaaaaaatagcaaccataaaaaggaatgaaagcatgtcctttgcagcaacatggatgcagctggaagccattatcctaa
  >chr1:1876909-1876911(+)
  aaatctgagttggcagaaccaggatGTCCAGCACGTGGATTCAGGTACAAGAAATACCACTGATATCTACATAATTACTACTCATGTTTTCCAAGATGAATG
  >chr1:2492296-2492298(-)
  AAAAATAAAGGAACGCCTCTCAGCTCGGGCCCTGCAAGAGCCACCTGCACAGGGCATGTTCCTCGAGCTCTCCTGCGGTGTCAGCTGTGCCGAGTGCTCCGA
  ```
  

* `donor_umi_counts.tsv` contains the UMI and read counts for reads that only map to the donor and not the junction. This can be useful as a proxy for integration efficiency/donor plasmid contamination.
  ```
  sample_id	donor_UMI	donor_RAW	donor_UNIQ
  sampleA	156037	165563	558
  sampleB	162923	234982	616
  sampleC	127094	196292	518
  ...
  ```
  * `sample_id` is the sample name as given in `metadata.tsv` 
  * `donor_UMI` is the number of UMIs that map only to the donor
  * `donor_RAW` is the number of unfiltered reads that align to the donor
  * `donor_UNIQ` is the number of uniquely mapped reads. 
  * Always use `donor_UMI` if the donor actually has UMIs.


* `merged_counts.sample.tsv` contains information about the number of reads per integration junction. This is information given on a per-sample basis.
* `merged_counts.biorep.tsv` is the same as `merged_counts.sample.tsv`, but merges the counts across technical replicates.
* `merged_counts.lsr.tsv` is the same as `merged_counts.biorep.tsv`, but merges the counts across all replicates for a given LSR.
* `raw_counts.unfiltered.tsv` is all insertion junction counts without any applied filters or merged insertion sites.
* `raw_counts.filtered.tsv` is all insertion junction counts with QC filters applied to insertion sites.
* `seqlogos/` is a directory that contains automatically generated sequence logos for the insertion sites of each sample.
* `seqlogos.txt` is just a log file that makes sure the sequence logo step ran successfully.